test/extensions/filters/http/wasm/wasm_filter_test.cc

```diff
@@ -501,6 +510,47 @@ TEST_P(WasmHttpFilterTest, BodyRequestBufferBody) {
   filter().onDestroy();
 }
 
+TEST_P(WasmHttpFilterTest, BodyRequestBufferBodyAndDestoryFilterInTheAddDecodedDataCallback) {
+  setupTest("body");
+  setupFilter();
+
+  Http::TestRequestHeaderMapImpl request_headers{{":path", "/"},
+                                                 {"x-test-operation", "BufferBody"}};
+  EXPECT_EQ(Http::FilterHeadersStatus::Continue, filter().decodeHeaders(request_headers, false));
+
+  Buffer::OwnedImpl bufferedBody;
+  EXPECT_CALL(decoder_callbacks_, decodingBuffer()).WillRepeatedly(Return(&bufferedBody));
+  EXPECT_CALL(decoder_callbacks_, modifyDecodingBuffer(_))
+      .WillRepeatedly(Invoke([&bufferedBody](BufferFunction f) { f(bufferedBody); }));
+
+  Buffer::OwnedImpl data1("hello");
+  EXPECT_CALL(filter(), log_(spdlog::level::err, Eq(absl::string_view("onBody hello"))));
+  EXPECT_EQ(Http::FilterDataStatus::StopIterationAndBuffer, filter().decodeData(data1, false));
+  // The first call to decodeData() will not result in the addDecodedData() callback being called
+  // because the filter doesn't know will the VM's onRequestBody() return StopIterationAndBuffer or
```

Comment:
"because the filter doesn't yet know if the VM's onRequestBody() will return StopIterationAndBuffer or Continue"